# Lab 2: Regression
# Part 2: Multiple Linear Regression

## Goals

The main goals of this labs is:
* Hands on python machine learning and data sciences packages (numpy, pandas, sklearn and matplotlib)
* Master of multiple linear regression algorithms practice. Miltiple linear rgression is a linear rgression depends on multiple features.
* compare simple and multiple linear regression algorithms performances using mse and r2_score.

## 1. Data Preparation

### 1.1. Importing and loading the Dataset

Use pandas for importing data from external data source
In our case we are working with the **house-prices.csv** file.

In [1]:
import pandas as pd

We specify the data source file url (from online source) or path, I recommand using absolute path rater than relative path to prevent FileNotFound Error, or just put the file under the same repository as the python script, and the file_url will be just "house-prices.csv"

In [2]:
file_url = "house-prices.csv"

We're now ready to import the CSV file into Python using read_csv() from pandas, with pandas you can read data from different source (csv, json, sql, xls/xlsx, ...) for more details you can see official pandas documentation

In [3]:
dataset = pd.read_csv(file_url)

We can now see our dataset table

In [4]:
# calling the head() function without passing an argument will return first 5 rows
# you can specify the number of rows in the head() method
# dataset.head(8)
dataset.head()

Home   Price  SqFt  Bedrooms  Bathrooms  Offers Brick Neighborhood
0     1  114300  1790         2          2       2    No         East
1     2  114200  2030         4          2       3    No         East
2     3  114800  1740         3          2       1    No         East
3     4   94700  1980         3          2       3    No         East
4     5  119800  2130         3          3       3    No         East

In [22]:
size = dataset[['SqFt']]
size # -> to see SqFt content

price = dataset[['Price']]
price # -> to see Price content

Price
0    114300
1    114200
2    114800
3     94700
4    119800
..      ...
123  119700
124  147900
125  113500
126  149900
127  124600

[128 rows x 1 columns]

Verifiying price and size vectors dimensions, then reshaping it (tranform from row vector to column vector).

In [23]:
import numpy as np

X = np.array(size).reshape(-1,1)
Y = np.array(price).reshape(-1,1)

# print(X)
# print(Y) 

The next step is to drop (delete) label column and all non-numerical data, **Brick** and **Neighborhood** columns.

In [11]:
data = dataset.drop(['Price', 'Brick', 'Neighborhood'], axis=1)
data.head()

Home  SqFt  Bedrooms  Bathrooms  Offers
0     1  1790         2          2       2
1     2  2030         4          2       3
2     3  1740         3          2       1
3     4  1980         3          2       3
4     5  2130         3          3       3

In [38]:
from sklearn.preprocessing import LabelEncoder

# create a Label Encoder object
le = LabelEncoder()

x1 = dataset['Brick']
new_x1 = le.fit_transform(x1)
new_x1 = new_x1.reshape(-1, 1)

x2 = dataset['Neighborhood']
new_x2 = le.fit_transform(x2)
new_x2 = new_x2.reshape(-1, 1)

# concatenate new_x1 and new_x2 to data array
new_data = np.concatenate((data, new_x1, new_x2), axis=1)

# covert new_data array to DataFrame
new_df = pd.DataFrame(new_data)
# set df columns names
new_df.columns = ['Home', 'SqFt', 'Bedrooms', 'Bathrooms', 'Offers', 'new_x1', 'new_x2']
new_df.head()

Home  SqFt  Bedrooms  Bathrooms  Offers  new_x1  new_x2
0     1  1790         2          2       2       0       0
1     2  2030         4          2       3       0       0
2     3  1740         3          2       1       0       0
3     4  1980         3          2       3       0       0
4     5  2130         3          3       3       0       0

## 2. Linear Regression

No we can use sklearn package to import *LinearRegression* from the linear_model module to create our linear regression model having the equation Y = a.X + b

In [39]:
from sklearn.linear_model import LinearRegression

# create the model
model = LinearRegression()

# fit (train) the model
model.fit(new_df, Y)

LinearRegression()

From this model we can now get the a and b coefficients.

In [40]:
a = model.coef_[0][0]
b = model.intercept_[0]
print("a = ", a)
print("b = ", b)

a =  -24.50627273642104
b =  -16118.943696606526


In [41]:
y_pred = model.predict(new_df)

You can see difference between the real prices vector Y, and the predicted prices vector y_pred.

In [42]:
for index in range(0, 5):
    print('For y = {}, \ty_pred = {}'.format(Y[[index]], y_pred[[index]]))

For y = [[114300]], 	y_pred = [[100032.58908853]]
For y = [[114200]], 	y_pred = [[116013.31884989]]
For y = [[114800]], 	y_pred = [[113588.49484655]]
For y = [[94700]], 	y_pred = [[106894.16588971]]
For y = [[119800]], 	y_pred = [[124726.28685475]]


#### Mean Squred Error

In [43]:
from sklearn.metrics import mean_squared_error
import math

mse = mean_squared_error(Y, y_pred)
math.sqrt(mse)

10240.343882718636

This value can not give a clear indicator for our model perfermance. we still need to compare this value to min, max and Y mean values.

In [44]:
# create a DataFrame from Y numpy array
ydf = pd.DataFrame(Y)

# use the describe() function to get mean, min, max and others details
ydf.describe()

0
count     128.000000
mean   130427.343750
std     26868.770371
min     69100.000000
25%    111325.000000
50%    125950.000000
75%    148250.000000
max    211200.000000

#### Using r2_score 

In [45]:
from sklearn.metrics import r2_score
r2_score(Y, y_pred)

0.8536004481639945

This model r2_score near 1, that's mean that this model having good performance.

### Conclusion

By comparing the r2_score of multiple linear regression (with value 0.85) to simple model r2_score calculated on part 1 of this lab (value 0.3) we can assume that **multiple linear regression produce a prediction with higher precision than simple linear regression**.